lib/sidekiq/api.rb

```diff
@@ -520,35 +526,41 @@ def find_job(jid)
     end
 
     def delete(score, jid = nil)
-      if jid
-        elements = Sidekiq.redis do |conn|
-          conn.zrangebyscore(name, score, score)
-        end
+      jid ? delete_by_jid(score, jid) : delete_by_score(score, score)
+    end
 
-        elements_with_jid = elements.map do |element|
-          message = Sidekiq.load_json(element)
+    def delete_by_value(name, value)
+      Sidekiq.redis do |conn|
+        ret = conn.zrem(name, value)
+        @_size -= 1 if ret
+        ret
+      end
+    end
 
-          if message["jid"] == jid
-            _, @_size = Sidekiq.redis do |conn|
-              conn.multi do
-                conn.zrem(name, element)
-                conn.zcard name
-              end
-            end
-          end
-        end
-        elements_with_jid.count != 0
-      else
-        count, @_size = Sidekiq.redis do |conn|
-          conn.multi do
-            conn.zremrangebyscore(name, score, score)
-            conn.zcard name
+    def delete_by_jid(score, jid)
+      elements = Sidekiq.redis do |conn|
+        conn.zrangebyscore(name, score, score)
+      end
+      elements.each do |element|
+        message = Sidekiq.load_json(element)
+        if message["jid"] == jid
+          break Sidekiq.redis do |conn|
+            ret = conn.zrem(name, element)
+            @_size -= 1 if ret
+            ret
           end
         end
-        count != 0
+        false
       end
     end
 
+    def delete_by_score(score_min, score_max)
```

Comment:
@mperham 
So currently Sidekiq will delete multiple entries if they are all scheduled to run at the same-time, but since score is a float unless one explicitly schedules things to do this I imagine it's quite difficult to do by mistake - it also appears to have been an intended behaviour https://github.com/pik/sidekiq/blob/master/test%2Ftest_api.rb#L343.

I can modify this if you would prefer to remove the old behaviour entirely - but if it is there imho there is no reason for not allowing greater versatility with a min/max_score option (and perhaps a comment urging caution when using this).